myia/opt/dde.py

```diff
@@ -1,226 +1,479 @@
-        the output of the graph. The empty path represents the
-        return value itself.
+        self.manager = resources.manager
+        # need :: node -> {paths}
+        self.need = defaultdict(set)
+        # flow :: node -> {node}
+        self.flow = defaultdict(set)
+        # backflow :: node -> {node} (opposite of flow)
+        self.backflow = defaultdict(set)
+        # results :: node -> need -> values
+        self.results = defaultdict(lambda: defaultdict(set))
+        self.todo = set()
+        self.run(root)
+
+    def run(self, root):
+        """Run the algorithm."""
+        for p in root.parameters:
+            self.add_value(p, WILDCARD, ANYTHING)
+        for ct in self.manager.all_nodes:
+            if ct.is_constant():
+                if isinstance(ct.value, tuple):
+                    self.add_value(ct, WILDCARD, ANYTHING)
+                else:
+                    self.add_value(ct, ANYTHING, ct.value)
+        self.add_need(root.return_, ANYTHING)
+        while self.todo:
+            nxt = self.todo.pop()
+            self.process_node(nxt)
+
+    def values(self, node, need):
+        """Return the current set of possible values for node:need."""
+        # The purpose of the WILDCARD key is to act as a fallback to cover any
+        # need. For example, if a parameter is a tree (so its type at this
+        # point is a recursive AbstractTuple), we might need paths (1, 0), (1,
+        # 1, 0), and so on, but rather than having an infinite number of
+        # entries, we just have a single WILDCARD entry that covers them all.
+        # This is conceptually different from ANYTHING, which covers the path
+        # (), where the need is for the unindexed data structure. WILDCARD
+        # is currently only set in run(), for parameters and tuple constants.
+        # So most of the time there's nothing in it.
+        return self.results[node][need] | self.results[node][WILDCARD]
+
+    def add_value(self, node, need, value):
+        """Propagate a value for a need on a node."""
+        if need is ANYTHING and isinstance(value, tuple):
+            for i, v in enumerate(value):
+                self.add_value(node, (i,), v)
+        self._add_value(node, need, value)
+
+    def _add_value(self, node, need, value):
+        assert need != ()
+        results = self.results[node][need]
+        if value not in results:
+            results.add(value)
+            # Invalidate all uses of this node, because they now have new
+            # values to test.
+            for node2, i in self.manager.uses[node]:
+                self.todo.add(node2)
+            for other_node in self.flow[node]:
+                self._add_value(other_node, need, value)
+
+    def add_need(self, node, need):
+        """Declare that node has the given need."""
+        needs = self.need[node]
+        if isinstance(need, tuple) and len(need) > MAX_NEED_DEPTH:
+            need = need[:MAX_NEED_DEPTH]
```

Comment:
I understand that this is here to prevent running forever, but I don't see any devolving mechanism to convert to a WILDCARD need or something.  That would be more prudent.

There might be something I don't get which alleviates this requirement.

Reply:
The need `(x, y)` subsumes the need `(x, y, z)` for all `z`, so it's safe to cut the tuple to any length.